aten/src/ATen/Error.h

```diff
@@ -25,262 +25,118 @@
-struct pack;
-template <bool... values>
-struct all_of : std::is_same<pack<values..., true>, pack<true, values...>> {};
+template <typename T, typename... Args>
+inline void
+_str(std::stringstream& ss, const T& t, const Args&... args) {
+  _str(ss, t);
+  _str(ss, args...);
+}
 
-/// A printf wrapper that returns an std::string.
-inline std::string format(const char* format_string, ...) {
-  static constexpr size_t kMaximumStringLength = 4096;
-  char buffer[kMaximumStringLength];
+} // namespace detail
 
-  va_list format_args;
-  va_start(format_args, format_string);
-  vsnprintf(buffer, sizeof(buffer), format_string, format_args);
-  va_end(format_args);
+// Convert a list of string-like arguments into a single string.
+template <typename... Args>
+inline std::string str(const Args&... args) {
+  std::stringstream ss;
+  detail::_str(ss, args...);
+  return ss.str();
+}
 
-  return buffer;
+// Specializations for already-a-string types.
+template <>
+inline std::string str(const std::string& str) {
+  return str;
+}
+inline std::string str(const char* c_str) {
+  return c_str;
 }
 
 /// Represents a location in source code (for debugging).
 struct SourceLocation {
-  std::string toString() const {
-    return format("%s at %s:%d", function, file, line);
-  }
-
   const char* function;
   const char* file;
   uint32_t line;
 };
-} // namespace detail
+
+inline std::ostream& operator<<(std::ostream& out, const SourceLocation& loc) {
+  out << loc.function << " at " << loc.file << ":" << loc.line;
+  return out;
+}
+
+std::string get_backtrace(
+    size_t frames_to_skip = 0,
+    size_t maximum_number_of_frames = 64);
 
 /// The primary ATen error class.
 /// Provides a complete error message with source location information via
 /// `what()`, and a more concise message via `what_without_backtrace()`. Should
 /// primarily be used with the `AT_ERROR` macro.
-struct AT_API Error : public std::exception {
-  template <typename... FormatArgs>
-  Error(
-      detail::SourceLocation source_location,
-      const char* format_string,
-      FormatArgs&&... format_args)
-      : what_without_backtrace_(detail::format(
-            format_string,
-            std::forward<FormatArgs>(format_args)...)),
-        what_(what_without_backtrace_) {
-    // NOTE: A "literal type"
-    // (http://en.cppreference.com/w/cpp/concept/LiteralType) could also be a
-    // constexpr struct, so it's not 100% guaranteed that the `printf` call
-    // inside `format` is safe, but it will catch 99.9% of all errors we'll run
-    // into, such as passing `std::string`.
-    static_assert(
-        detail::all_of<std::is_literal_type<FormatArgs>::value...>::value,
-        "format arguments must be literal types!");
-    what_ += " (" + source_location.toString() + ")\n";
-#if !defined(_WIN32)
-    // Skip this constructor's frame.
-    what_ += detail::get_backtrace(/*frames_to_skip=*/1);
-#endif // !defined(_WIN32)
-  }
+class AT_API Error : public std::exception {
```

Comment:
people tell me to use `struct` 🤷‍♂️

Reply:
I referred to the PR which introduced this code https://github.com/pytorch/pytorch/pull/6059 but I didn't see any direction that `struct` should be a struct. I also searched for "std::exception class struct" and didn't see any guidance. What's the motivation for the rule here?

In this case, I flipped struct into a class because I saw there were private members and there was non-trivial functionality for constructing the data. (see also https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes )

Reply:
It's not my rule (I'm totally with you), but when I refactored `Variable` and `Function` in the autograd I was told that we prefer `struct`. It's not good C++ style IMHO, but most classes in ATen and autograd use struct.

Reply:
I remember that being one of the reasons yes. Although I really like the mental distinction between a `class` being an abstraction that has behavior, and a `struct` being a simple POD type

Reply:
Yes, I've been arguing for using `struct` everywhere because ultimately that makes things much more consistent, and lowers the mental overhead.

#### \<rant>

C++ does a reasonably good job of hiding the `struct` vs `class` for you, except when you need to forward declare. This is especially annoying because not all compilers treat it as a warning, so the author of a commit might get a nice clean build, while others will get a big log. 

Arguing that `class` should be for non-trivial functionality, while `struct` is for POD is also not great, because this essentially means that all use sites need to know about it's implementation, which violates the whole OOP idea (implementation != interface!). Also, the POD rule really doesn't make the distinction easy, because the struct below is a POD
```cpp
struct MyStruct1 {
    int x;
    void print_x() {
        std::cout << x << std::endl;
    }
    // <insert 100 more methods here>
};
```
while this one isn't
```cpp
struct MyStruct2 {
    int x = 2;
};
```
([godbolt link for reference](https://godbolt.org/g/NxnJ55)). Additionally, what happens if a struct is a trivial POD, but then you add a new member with a default initializer? Do you have to go around and change all the forward declarations to match the new type? That's just counterproductive.

#### \</rant>

Having said this, if we're planning to migrate this code to C10, and C10 is going to use a different coding style then we can of course start changing it in here.